### DISCUSSION TITLE
Inconsistent Prisma performance on scaling the number of concurrent operations

### DISCUSSION THREAD

**Discussion by srujun** [prisma/prisma#23388] [Q&A]
### Question

I am running a GraphQL backend in Node.js with Apollo, and using Prisma as the ORM layer on top of Postgres. The server is running in GCP on Cloud Run, and I use a managed Postgres instance through GCP's Cloud SQL.

I am seeing very inconsistent performance with Prisma, especially in some of the more complex GraphQL resolvers which are resolving multiple database entities, and this problem gets worse as the number of concurrent queries increases. I have implemented Prisma's tracing feature with OpenTelemetry to investigate this problem, and there are 3 main issues I am running into.

## Issue 1: expensive `SELECT 1` query when waiting for connection

I am using `?connection_limit=40` in my Postgres connection string, which should leave enough connections available for Prisma. I am noticing that sometimes there are random instances of the `prisma:engine:connection` span which lasts anywhere from 700 ms to 1 sec. It is unclear why this query should take so long when most of the other queries are finishing in within 50 ms to 100 ms.

![Issue 1](https://i.stack.imgur.com/iRb9H.png)

## Issue 2: individual queries take longer as the number of concurrent queries increases

I am noticing that as I am scaling up the number of concurrent database queries within the same GraphQL query, the same individual queries take longer to finish. Generally most queries finish within 50 ms to 100 ms when there are fewer concurrent queries, but the same queries in more expensive GraphQL resolvers are running for almost 300 ms.

**Few queries:**

![Issue 2: few queries](https://i.stack.imgur.com/PH1DL.png)

**More queries:**

![Issue 2: more queries](https://i.stack.imgur.com/ZHRP8.png)

The Postgres instance I am using has 2 vCPUs and 8 GB memory, and I have checked all the charts on the Cloud SQL console. The CPU utilization is under 10% consistently, number of active connections peaks at ~100 whereas the limit is 400. I have also looked at some of the data in Query Insights and there is nothing that stands out in terms of query performance, so it definitely looks like this issue is happening somewhere in Prisma or how I am configuring Prisma.

## Issue 3: Prisma telemetry spans are missing

There are many `prisma:client:operation` spans that don't have anything other than a `prisma:client:serialize` child span. It is unclear why these spans are missing unless there is some DataLoader optimization happening within which is not instrumented, or that this is an issue in Prisma's tracing feature?

![Issue 3](https://i.stack.imgur.com/SE3Z7.png)

### Information about Prisma Schema, Client Queries and Environment (optional)

- Database: PostgreSQL 12.17
- Node.js version: 18.16

```
prisma -v
Environment variables loaded from .env
prisma                  : 5.10.2
@prisma/client          : 5.10.2
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v18.12.1
Query Engine (Node-API) : libquery-engine 5a9203d0590c951969e85a7d07215503f4672eb9 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 5a9203d0590c951969e85a7d07215503f4672eb9 (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.10.0-34.5a9203d0590c951969e85a7d07215503f4672eb9
Default Engines Hash    : 5a9203d0590c951969e85a7d07215503f4672eb9
Studio                  : 0.499.0
Preview Features        : tracing
```

**Comment by ludralph** [comment]
Hi @srujun 👋 

Thank you for raising this question.

> Issue 1: Expensive SELECT 1 query when waiting for connection

Can you confirm that the query is actually recorded as that slow on the database side as well? The tracing preview feature may be broken and reporting the wrong duration.

> Issue 2: Individual queries take longer as the number of concurrent queries increases

One potential solution is to optimize your queries by improving the performance of your queries by adding indexes where necessary. 

> Issue 3: Prisma telemetry spans are missing

This could potentially be a bug in Prisma's tracing feature as it is currently in Preview.

**Reply by AlexMog** [reply]
TL;DR: I can confirm that database queries are optimized on my side, and that Prisma seems to be consuming a lot of CPU.

Setup: Nestjs graphql + Prisma

Load for the usage of 16x3 threads (on 3 servers): 30 req/s (which is really low, I can dos my own cluster with a single endpoint spam, even by spamming a simple query)

I am having the exact same problem, the CPU spikes to 100% with less than 30req/s (on a 3 server 8cx ores 16T setup (3.4 GHz per core)), the db queries take less than 5ms to answer, my CPU seems to be used by Prisma, I have the exact same setup for another project (same framework, grapql (through nestjs) but uses only a Redis db, I see no spikes on the CPU side, it can handle 10k+ graphql complex queries /s on the same setup)

It seems to be a Prisma related problem, I have two preview features enabled: joins and tracing, might it be related ?

+ I had to replace all my "find unique" with data loaders, the documentation claims that find unique is perfect for resolvers because it batches on node ticks, but the CPU performance is horrible, gained 300% performance just by replacing the Prisma.myColumn.findUnique({ id }).myJoin() with a find many if the myJoin column through dataloaders


We are also experiencing the same problem on another project which uses a rest endpoint with a single find unique.
Cap at 20req/s on 6 cores (aws vcores graviton series on 6 virtual servers)
Same preview features enabled, so it does not seems to be related to graphql, but indeed a problem with Prisma CPU usage (we observe a 100% CPU usage on this project too, with queries taking less than 1ms on the db side)

**Reply by ludralph** [reply]
Hi @AlexMog 

Do you still experience this issue in the latest release of Prisma 6.2.1? Tracing has since gone to GA. You can check the [release note](https://github.com/prisma/prisma/releases/tag/6.1.0).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
